File: src/main/java/io/github/toberocat/guiengine/api/utils/VirtualInventory.java
package io.github.toberocat.guiengine.api.utils;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.HumanEntity;
import org.bukkit.event.inventory.InventoryType;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.InventoryHolder;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.ListIterator;

/**
 * Created: 29.04.2023
 *
 * @author Tobias Madlberger (Tobias)
 */
public class VirtualInventory implements Inventory {
    private final int height;
    private final Runnable reload;

    public VirtualInventory(int height, Runnable reload) {
        this.height = height;
        this.reload = reload;
    }

    @Override
    public int getSize() {
        return height * 9;
    }

    @Override
    public int getMaxStackSize() {
        return 64;
    }

    @Override
    public void setMaxStackSize(int size) {
    }

    @Nullable
    @Override
    public ItemStack getItem(int index) {
        return null;
    }

    @Override
    public void setItem(int index, @Nullable ItemStack item) {

    }

    @NotNull
    @Override
    public HashMap<Integer, ItemStack> addItem(@NotNull ItemStack... items) throws IllegalArgumentException {
        return new HashMap<>();
    }

    @NotNull
    @Override
    public HashMap<Integer, ItemStack> removeItem(@NotNull ItemStack... items) throws IllegalArgumentException {
        return new HashMap<>();
    }

    @NotNull
    @Override
    public ItemStack[] getContents() {
        return new ItemStack[54];
    }

    @Override
    public void setContents(@NotNull ItemStack[] items) throws IllegalArgumentException {
        reload.run(); // ToDo: Maybe make this one only reload own inventory
    }

    @NotNull
    @Override
    public ItemStack[] getStorageContents() {
        return new ItemStack[0];
    }

    @Override
    public void setStorageContents(@NotNull ItemStack[] items) throws IllegalArgumentException {

    }

    @Override
    public boolean contains(@NotNull Material material) throws IllegalArgumentException {
        return false;
    }

    @Override
    public boolean contains(@Nullable ItemStack item) {
        return false;
    }

    @Override
    public boolean contains(@NotNull Material material, int amount) throws IllegalArgumentException {
        return false;
    }

    @Override
    public boolean contains(@Nullable ItemStack item, int amount) {
        return false;
    }

    @Override
    public boolean containsAtLeast(@Nullable ItemStack item, int amount) {
        return false;
    }

    @NotNull
    @Override
    public HashMap<Integer, ? extends ItemStack> all(@NotNull Material material) throws IllegalArgumentException {
        return new HashMap<>();
    }

    @NotNull
    @Override
    public HashMap<Integer, ? extends ItemStack> all(@Nullable ItemStack item) {
        return new HashMap<>();
    }

    @Override
    public int first(@NotNull Material material) throws IllegalArgumentException {
        return 0;
    }

    @Override
    public int first(@NotNull ItemStack item) {
        return 0;
    }

    @Override
    public int firstEmpty() {
        return 0;
    }

    @Override
    public void remove(@NotNull Material material) throws IllegalArgumentException {

    }

    @Override
    public void remove(@NotNull ItemStack item) {

    }

    @Override
    public void clear(int index) {

    }

    @Override
    public void clear() {

    }

    @NotNull
    @Override
    public List<HumanEntity> getViewers() {
        return new ArrayList<>();
    }

    @NotNull
    @Override
    public InventoryType getType() {
        return InventoryType.CHEST;
    }

    @Nullable
    @Override
    public InventoryHolder getHolder() {
        return null;
    }

    @NotNull
    @Override
    public ListIterator<ItemStack> iterator() {
        return new ArrayList<ItemStack>().listIterator();
    }

    @NotNull
    @Override
    public ListIterator<ItemStack> iterator(int index) {
        return new ArrayList<ItemStack>().listIterator();
    }

    @Nullable
    @Override
    public Location getLocation() {
        return null;
    }
}

File: src/main/kotlin/io/github/toberocat/guiengine/utils/VirtualInventory.kt
package io.github.toberocat.guiengine.utils

import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.HumanEntity
import org.bukkit.event.inventory.InventoryType
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.InventoryHolder
import org.bukkit.inventory.ItemStack

/**
 * Created: 29.04.2023
 *
 * @author Tobias Madlberger (Tobias)
 */
class VirtualInventory(private val height: Int, private val reload: Runnable) : Inventory {
    override fun getSize(): Int {
        return height * 9
    }

    override fun getMaxStackSize(): Int {
        return 64
    }

    override fun setMaxStackSize(size: Int) {}
    override fun getItem(index: Int): ItemStack? {
        return null
    }

    override fun setItem(index: Int, item: ItemStack?) {}

    @Throws(IllegalArgumentException::class)
    override fun addItem(vararg items: ItemStack): HashMap<Int, ItemStack> {
        return HashMap()
    }

    @Throws(IllegalArgumentException::class)
    override fun removeItem(vararg items: ItemStack): HashMap<Int, ItemStack> {
        return HashMap()
    }

    override fun getContents(): Array<ItemStack> {
        return Array(54) { ItemStack(Material.AIR) }
    }

    @Throws(IllegalArgumentException::class)
    override fun setContents(items: Array<ItemStack>) {
        reload.run() // ToDo: Maybe make this one only reload own inventory
    }

    override fun getStorageContents(): Array<ItemStack> {
        return Array(0) { ItemStack(Material.AIR) }
    }

    @Throws(IllegalArgumentException::class)
    override fun setStorageContents(items: Array<ItemStack>) {
    }

    @Throws(IllegalArgumentException::class)
    override fun contains(material: Material): Boolean {
        return false
    }

    override fun contains(item: ItemStack?): Boolean {
        return false
    }

    @Throws(IllegalArgumentException::class)
    override fun contains(material: Material, amount: Int): Boolean {
        return false
    }

    override fun contains(item: ItemStack?, amount: Int): Boolean {
        return false
    }

    override fun containsAtLeast(item: ItemStack?, amount: Int): Boolean {
        return false
    }

    @Throws(IllegalArgumentException::class)
    override fun all(material: Material): HashMap<Int, out ItemStack> {
        return HashMap()
    }

    override fun all(item: ItemStack?): HashMap<Int, out ItemStack> {
        return HashMap()
    }

    @Throws(IllegalArgumentException::class)
    override fun first(material: Material): Int {
        return 0
    }

    override fun first(item: ItemStack): Int {
        return 0
    }

    override fun firstEmpty(): Int {
        return 0
    }

    override fun isEmpty(): Boolean {
        return false
    }

    @Throws(IllegalArgumentException::class)
    override fun remove(material: Material) {
    }

    override fun remove(item: ItemStack) {}
    override fun clear(index: Int) {}
    override fun clear() {}
    override fun getViewers(): List<HumanEntity> {
        return ArrayList()
    }

    override fun getType(): InventoryType {
        return InventoryType.CHEST
    }

    override fun getHolder(): InventoryHolder? {
        return null
    }

    override fun iterator(): MutableListIterator<ItemStack> {
        return ArrayList<ItemStack>().listIterator()
    }

    override fun iterator(index: Int): ListIterator<ItemStack> {
        return ArrayList<ItemStack>().listIterator()
    }

    override fun getLocation(): Location? {
        return null
    }
}
